examples/layers/services/isolate.dart

```diff
@@ -0,0 +1,308 @@
+    JsonDecoder decoder = new JsonDecoder(
+      (dynamic key, dynamic value) {
+        if (onProgressListener != null) {
+          if (key is int && i++ % _NOTIFY_INTERVAL == 0)
+            onProgressListener(i.toDouble(), _NUM_ITEMS.toDouble());
+        }
+        return value;
+      }
+    );
+    try {
+      List<dynamic> result = decoder.convert(_data);
+      int n = result.length;
+      if (onResultListener != null)
+        onResultListener("Decoded $n results");
+    } catch (e, stack) {
+      print("Invalid JSON file: $e");
+      print(stack);
+    }
+  }
+
+  static String _replicateJson(String data, int count) {
+    StringBuffer buffer = new StringBuffer()..write("[");
+    for (int i = 0; i < count; i++) {
+      buffer.write(data);
+      if (i < count - 1)
+        buffer.write(',');
+    }
+    buffer.write("]");
+    return buffer.toString();
+  }
+}
+
+/// The current state of the calculation.
+enum CalculationState {
+  idle,
+  loading,
+  calculating
+}
+
+/// A manager for the connection to a spawned isolate.
+///
+/// Isolates communicate with each other via [ReceivePort]s and [SendPort]s.
+/// This class manages these ports and maintains state related to the
+/// progress of the background computation.
+class CalculationManager {
+  CalculationManager({ this.onProgressListener, this.onResultListener })
+  : _receivePort = new ReceivePort() {
+    _receivePort.listen(_handleMessage);
+  }
+
+  CalculationState _state = CalculationState.idle;
+  CalculationState get state => _state;
+  bool get isRunning => _state != CalculationState.idle;
+
+  double _completed = 0.0;
+  double _total = 1.0;
+
+  final OnProgressListener onProgressListener;
+  final OnResultListener onResultListener;
+
+  /// Start the background computation.
+  ///
+  /// Does nothing if the computation is already running.
+  void start() {
+    if (!isRunning) {
+      _state = CalculationState.loading;
+      _runCalculation();
+    }
+  }
+
+  /// Stop the background computation.
+  ///
+  /// Kills the [Isolate] immediately, if spawned. Does nothing if the
+  /// computation is not running.
+  void stop() {
+    if (isRunning) {
+      _state = CalculationState.idle;
+      if (_isolate != null) {
+        _isolate.kill(priority: Isolate.IMMEDIATE);
+        _isolate = null;
+        _completed = 0.0;
+        _total = 1.0;
+      }
+    }
+  }
+
+  final ReceivePort _receivePort;
+  Isolate _isolate;
+
+  void _runCalculation() {
+    // Load the JSON string. Note that this is done in the main isolate; at the
+    // moment, spawned isolates do not have access to Mojo services, including
+    // the root bundle (see https://github.com/flutter/flutter/issues/3294).
+    // However, the loading process is asynchronous, so the UI will not block
+    // while the file is loaded.
+    rootBundle.loadString('services/data.json').then((String data) {
+      if (isRunning) {
+        List<dynamic> message = [ _receivePort.sendPort, data ];
```

Comment:
I believe this is governed by the rules here: https://api.dartlang.org/stable/1.15.0/dart-isolate/SendPort/send.html.

It turns out I was holding it wrong though. Will fix to use a structured message.